File: java/src/main/java/com/fleetmgr/sdk/client/state/device/connected/Flying.java
package com.fleetmgr.sdk.client.state.device.connected;

import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.connection.Received;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.traffic.Channel;
import com.fleetmgr.sdk.client.event.output.facade.ChannelsClosed;
import com.fleetmgr.sdk.client.event.output.facade.ChannelsOpened;
import com.fleetmgr.sdk.client.event.output.facade.FacadeEvent;
import com.fleetmgr.sdk.client.state.State;
import com.fleetmgr.interfaces.ChannelsResponse;
import com.fleetmgr.interfaces.facade.control.ClientMessage;
import com.fleetmgr.interfaces.facade.control.Command;
import com.fleetmgr.interfaces.facade.control.ControlMessage;
import com.fleetmgr.interfaces.facade.control.Response;

import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Flying extends State {

    private List<com.fleetmgr.interfaces.Channel> initialChannels;

    Flying(State state, List<com.fleetmgr.interfaces.Channel> initialChannels) {
        super(state);
        this.initialChannels = initialChannels;
    }

    @Override
    public State start() {
        attachChannels(initialChannels);
        return null;
    }

    @Override
    public State notifyEvent(UserEvent event) {
        switch (event.getType()) {
            default:
                return defaultEventHandle(event.toString());
        }
    }

    @Override
    public State notifyConnection(ConnectionEvent event) {
        switch (event.getType()) {
            case RECEIVED:
                return handleMessage(((Received)event).getMessage());

            case LOST:
                return new Recovering(this);

            default:
                return defaultEventHandle(event.toString());
        }
    }

    private State handleMessage(ControlMessage message) {
        switch (message.getCommand()) {
            case ATTACH_CHANNELS:
                attachChannels(
                        message.getAttachChannels().getChannelsList());
                return null;

            case RELEASE_CHANNELS:
                releaseChannels(message.getChannels().getAttachedChannelsList());
                return null;

            case OPERATION_ENDED:
                Collection<Long> openedChannels = backend.getChannelsHandler().getChannelIds();
                listener.onEvent(new ChannelsClosed(openedChannels));
                backend.getChannelsHandler().closeAllChannels();
                listener.onEvent(new FacadeEvent(FacadeEvent.Type.OPERATION_ENDED));
                send(ClientMessage.newBuilder()
                        .setCommand(Command.OPERATION_ENDED)
                        .setResponse(Response.ACCEPTED)
                        .setAttachChannels(ChannelsResponse.newBuilder()
                                .addAllAttachedChannels(openedChannels)
                                .build())
                        .build());
                return new Ready(this);

            default:
                return defaultMessageHandle(message);
        }
    }

    private void attachChannels(List<com.fleetmgr.interfaces.Channel> channels) {
        Map<Long, Channel> validated = backend.getChannelsHandler().validateChannels(channels);
        listener.onEvent(new ChannelsOpened(validated.values()));
        send(ClientMessage.newBuilder()
                .setCommand(Command.ATTACH_CHANNELS)
                .setResponse(Response.ACCEPTED)
                .setAttachChannels(ChannelsResponse.newBuilder()
                        .addAllAttachedChannels(validated.keySet())
                        .build())
                .build());
    }

    private void releaseChannels(List<Long> channels) {
        listener.onEvent(new ChannelsClosed(channels));
        backend.getChannelsHandler().closeChannels(channels);
        send(ClientMessage.newBuilder()
                .setCommand(Command.RELEASE_CHANNELS)
                .setResponse(Response.ACCEPTED)
                .setAttachChannels(ChannelsResponse.newBuilder()
                        .addAllAttachedChannels(channels)
                        .build())
                .build());
    }

    @Override
    public String toString() {
        return "Flying";
    }
}

File: cpp/source/state/device/connected/Flying.cpp
#include "state/device/connected/Flying.hpp"

#include "state/device/connected/Ready.hpp"

#include "event/input/connection/Received.hpp"

#include "event/output/ChannelsOpened.hpp"
#include "event/output/ChannelsClosing.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::device;
using namespace fm::state::device::connected;

using namespace com::fleetmgr::interfaces::facade::control;

using event::input::user::UserEvent;
using event::input::connection::ConnectionEvent;
using event::input::connection::Received;

using event::output::FacadeEvent;
using event::output::ChannelsOpened;
using event::output::ChannelsClosing;

using com::fleetmgr::interfaces::ChannelResponse;

Flying::Flying(IState& state, std::shared_ptr<std::vector<ChannelResponse>> _initialChannels) :
    IState(state),
    initialChannels(_initialChannels)
{
}

std::unique_ptr<IState> Flying::start()
{
    attachChannels(initialChannels);
    return nullptr;
}

std::string Flying::toString() const
{
    return "Flying";
}

std::unique_ptr<IState> Flying::handleUserEvent(const UserEvent& event)
{
    switch (event.getType())
    {
    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Flying::handleConnectionEvent(const ConnectionEvent& event)
{
    switch (event.getType())
    {
    case ConnectionEvent::RECEIVED:
        return handleMessage(reinterpret_cast<const Received&>(event).getMessage());

    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Flying::handleMessage(const ControlMessage& message)
{
    switch (message.command())
    {
    case Command::ATTACH_CHANNELS:
    {
        std::shared_ptr<std::vector<ChannelResponse>> openedeChannels =
                std::make_shared<std::vector<ChannelResponse>>();
        openedeChannels->reserve(message.channelsresponse().channels_size());
        for (int i = 0; i < message.channelsresponse().channels_size(); ++i)
        {
            openedeChannels->emplace_back(message.channelsresponse().channels(i));
        }
        attachChannels(openedeChannels);
        return nullptr;
    }

    case Command::RELEASE_CHANNELS:
    {
        std::shared_ptr<std::vector<long>> releasedChannels =
                std::make_shared<std::vector<long>>();
        releasedChannels->reserve(message.channelsindication().ids_size());
        for (int i = 0; i < message.channelsindication().ids_size(); ++i)
        {
            releasedChannels->emplace_back(message.channelsindication().ids(i));
        }
        releaseChannels(releasedChannels);
        return nullptr;
    }

    case Command::OPERATION_ENDED:
    {
        std::shared_ptr<std::vector<traffic::IChannel*>> channels =
                backend.getChannelsHandler().getChannels();
        listener.onEvent(std::make_shared<ChannelsClosing>(channels));
        backend.getChannelsHandler().closeAllChannels();
        listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::OPERATION_ENDED));
        ClientMessage response;
        response.set_command(Command::OPERATION_ENDED);
        response.set_response(Response::ACCEPTED);
        send(response);
        return std::make_unique<Ready>(*this);
    }

    default:
        return defaultMessageHandle(message);
    }
}

void Flying::attachChannels(std::shared_ptr<std::vector<ChannelResponse>> channels)
{
    std::shared_ptr<std::vector<traffic::IChannel*>> result =
            backend.getChannelsHandler().validateChannels(*channels);
    listener.onEvent(std::make_shared<ChannelsOpened>(result));
    ClientMessage response;
    response.set_command(Command::ATTACH_CHANNELS);
    response.set_response(Response::ACCEPTED);
    for (traffic::IChannel* c : *result)
    {
        response.mutable_channelsindication()->add_ids(c->getId());
    }
    send(response);
}

void Flying::releaseChannels(std::shared_ptr<std::vector<long>> channels)
{
    listener.onEvent(std::make_shared<ChannelsClosing>(
                         backend.getChannelsHandler().getChannels(*channels)));
    backend.getChannelsHandler().closeChannels(*channels);
    ClientMessage response;
    response.set_command(Command::RELEASE_CHANNELS);
    response.set_response(Response::ACCEPTED);
    for (long id : *channels)
    {
        response.mutable_channelsindication()->add_ids(id);
    }
    send(response);
}
